src/adapter/src/catalog/open/builtin_item_migration.rs

```diff
@@ -0,0 +1,540 @@
+                }
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    // 3. Read in the current contents of the migration shard.
+    let upper = fetch_upper(&mut write_handle).await;
+    let as_of = Antichain::from_elem(upper.saturating_sub(1));
+    let since = read_handle.since();
+    as_of.join(since);
+    let snapshot = read_handle
+        .snapshot_and_fetch(as_of)
+        .await
+        .expect("we have advanced the as_of by the since");
+    soft_assert_or_log!(
+        snapshot.iter().all(|(_, _, diff)| *diff == 1),
+        "snapshot_and_fetch guarantees a consolidated result: {snapshot:?}"
+    );
+    let mut global_id_shards: BTreeMap<_, _> = snapshot
+        .into_iter()
+        .map(|((key, value), _ts, _diff)| {
+            let table_key = key.expect("persist decoding error");
+            let shard_id = value.expect("persist decoding error");
+            (table_key, shard_id)
+        })
+        .collect();
+
+    // 4. Clean up contents of migration shard.
+    let mut migrated_shard_updates: Vec<((TableKey, ShardId), Timestamp, Diff)> = Vec::new();
+    let mut migration_shards_to_finalize = BTreeSet::new();
+    let storage_collection_metadata = {
+        let txn: &mut dyn StorageTxn<Timestamp> = txn;
+        txn.get_collection_metadata()
+    };
+    for (table_key, shard_id) in global_id_shards.clone() {
+        if table_key.deploy_generation > deploy_generation {
+            halt!(
+                "saw deploy generation {}, which is greater than current deploy generation {}",
+                table_key.deploy_generation,
+                deploy_generation
+            );
+        }
+
+        if !migrated_tables.contains(&table_key.global_id)
+            || table_key.deploy_generation < deploy_generation
+        {
+            global_id_shards.remove(&table_key);
+            if storage_collection_metadata.get(&GlobalId::System(table_key.global_id))
+                == Some(&shard_id)
+            {
+                migrated_shard_updates.push(((table_key, shard_id.clone()), upper, -1));
+            } else {
+                migration_shards_to_finalize.insert((table_key, shard_id));
+            }
+        }
+    }
+
+    // 5. Add migrated tables to migration shard for current generation.
+    let mut global_id_shards: BTreeMap<_, _> = global_id_shards
+        .into_iter()
+        .map(|(table_key, shard_id)| (table_key.global_id, shard_id))
+        .collect();
+    for global_id in migrated_tables {
+        if !global_id_shards.contains_key(&global_id) {
+            let shard_id = ShardId::new();
+            global_id_shards.insert(global_id, shard_id);
+            let table_key = TableKey {
+                global_id,
+                deploy_generation,
+            };
+            migrated_shard_updates.push(((table_key, shard_id), upper, 1));
+        }
+    }
+
+    write_to_migration_shard(
```

Comment:
Maybe comment here or on the method that it's very important that we write at the next available timestamp after the `upper` at which we read. Making sure that no-one else snuck in any updates.

Reply:
Done.

Reply:
+1, I was going to leave the same comment 👯